packages/client/src/pages/Game/maps/Map.ts

```diff
@@ -93,41 +255,91 @@ class Map {
-    );
-    // turn to the right
-    this.engine.mapContext?.rect(
-      this.mapParams.width / 2,
-      this.mapParams.bottomBorder + this.mapParams.gridStep,
-      this.mapParams.width / 2,
-      this.mapParams.gridStep,
-    );
-    this.engine.mapContext?.fill();
-    this.engine.mapContext?.closePath();
+    // draw map 2d representation due to map stages
+    this.stageArr.forEach((stage, index) => {
+      if (stage.direction !== "start" && stage.direction !== "end") {
+        this.engine.context!.map!.beginPath();
+        this.engine.context!.map!.fillStyle = this.mapParams.backgroundColor;
+        // draw map path according to stage array
+        switch (stage.direction) {
+          // right
+          case "right": {
+            this.engine.context!.map!.rect(
+              this.stageArr[index - 1].limit.x,
+              this.stageArr[index - 1].limit.y -
+                (this.stageArr[index - 1].direction === "start"
+                  ? 0
+                  : this.mapParams.gridStep),
+              stage.limit.x -
+                this.stageArr[index - 1].limit.x +
+                (this.stageArr[index - 1].direction === "start"
+                  ? 0
+                  : this.mapParams.gridStep) -
+                this.mapParams.gridStep,
+              this.mapParams.gridStep,
+            );
+            break;
+          }
+          // down
+          case "down": {
+            this.engine.context!.map!.rect(
+              this.stageArr[index - 1].limit.x -
+                (this.stageArr[index - 1].direction === "right"
+                  ? this.mapParams.gridStep
+                  : 0),
+              this.stageArr[index - 1].limit.y,
+              this.mapParams.gridStep,
+              stage.limit.y -
+                this.stageArr[index - 1].limit.y -
+                this.mapParams.gridStep +
+                (this.stageArr[index - 1].direction === "right"
+                  ? this.mapParams.gridStep
+                  : 0) +
+                (this.stageArr[index - 1].direction === "left"
+                  ? this.mapParams.gridStep
+                  : 0),
+            );
+            break;
+          }
+          // left
+          case "left": {
+            this.engine.context!.map!.rect(
+              stage.limit.x,
+              this.stageArr[index - 1].limit.y - this.mapParams.gridStep,
+              Math.abs(this.stageArr[index - 1].limit.x - stage.limit.x),
+              this.mapParams.gridStep,
+            );
+            break;
+          }
+          // up
+          case "up": {
+            this.engine.context!.map!.rect(
+              stage.limit.x - this.mapParams.gridStep,
+              stage.limit.y - this.mapParams.gridStep,
+              this.mapParams.gridStep,
+              this.stageArr[index - 1].limit.y - stage.limit.y,
+            );
+            break;
+          }
+          default: {
+            throw new Error(`Unknown direction type: ${stage.direction}`);
```

Comment:
Nice to see that unexpected behavior handled 👍🏻

Reply:
First and the only try..catch :D